blog/en/blog/2022/08/28/fault-injection-testing-with-api-gateway.md

```diff
@@ -0,0 +1,202 @@
+
+To test the shopping application’s microservices for resiliency, we are going to simulate the product service misbehaving as a faulty service:
+
+- By adding a **delay** to the HTTP request.
+- By **aborting** the HTTP requests and returning a custom status code.
+
+### Injecting an HTTP delay fault
+
+In the first example, we introduce a 5-second delay for every request to the product service to test if we correctly set a connection timeout for calls to the product service from the Catalog service.
+
+> Note that you can also specify the percentage of requests to be delayed in numbers. Like 10 means: _10% of overall requests will be delayed_. In our demo case, we made it 100% to easily test the delay in time.
+
+The following route configuration example creates a new upstream for our backend service (productapi) that runs on port `80`, and registers a route with the `fault-injection` plugin enabled. You can notice that we set the delay injection in the plugin settings:
+
+``` bash
+curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+  "name": "Route for Fault Injection with the delay",
+  "methods": ["GET"],
+  "uri": "/api/products",
+  "plugins": {
+     "fault-injection": {
+         "delay": {
+            "duration": 5,
+   "percentage": 10
+         }
+     }
+  },
+ "upstream": {
+ "type": "roundrobin",
+ "nodes": {
+ "productapi:80": 1
+  }
+ }
+}'
+```
+
+Below we confirm the rule was created by running another `curl` command with the time measurement:
+
+``` bash
+time curl http://127.0.0.1:9080/api/products -i
+```
+
+After you run the cmd, you will see there is some delay was introduced:
+
+``` text
+HTTP/1.1 200 OK
+Content-Type: application/json; charset=utf-8
+Transfer-Encoding: chunked
+Connection: keep-alive
+
+[{"name":"Macbook Pro","price":1500.9},{"name":"SurfaceBook 3","price":1599.9}]
+real    0m5.004s
+user    0m0.004s
+sys     0m0.000s
+```
+
+The result of fault injection is as we expected.👏
+
+### Injecting an HTTP abort fault
+
+In the following example, we will introduce an HTTP abort to the product microservice to check how our imaginary Catalog service responds immediately to the failures introduced by the dependent service. Let’s say when the Product service fails, we should expect an HTTP error with the `Product service currently unavailable` error message.
+
+We can test it in action. Now we can enable abort injection with the following route settings.
+
+``` bash
+curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+  "name": "Route for Fault Injection with the abort",
+  "methods": ["GET"],
+  "uri": "/api/products",
+  "plugins": {
+     "fault-injection": {
+         "abort": {
+            "http_status": 503,
+   "body": "The product service is currently unavailable.",
+   "percentage": 100
+         }
+     }
+  },
+ "upstream": {
+ "type": "roundrobin",
```

Comment:
Hi Bobur, could you please format the json?

Reply:
Sure, I did!